This lecture is adapted from the great 'Introduction to Python' course from Eric Matthes (http://introtopython.org/) by Fabian Flöck and Florian Lemmerich and, like the original, is available under an [MIT license](#License). 

Variables, Strings, Numbers and Operators
===

In this section, you will learn to store information in variables. You will learn about two types of data: strings, which are sets of characters, and numerical data types.

# Table of Contents
 <p><div class="lev1 toc-item"><a href="#Variables,-Strings,-Numbers-and-Operators" data-toc-modified-id="Variables,-Strings,-Numbers-and-Operators-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Variables, Strings, Numbers and Operators</a></div><div class="lev1 toc-item"><a href="#Variables" data-toc-modified-id="Variables-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Variables</a></div><div class="lev2 toc-item"><a href="#Example" data-toc-modified-id="Example-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Example</a></div><div class="lev2 toc-item"><a href="#Naming-rules" data-toc-modified-id="Naming-rules-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Naming rules</a></div><div class="lev2 toc-item"><a href="#NameError" data-toc-modified-id="NameError-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>NameError</a></div><div class="lev2 toc-item"><a href="#Exercises" data-toc-modified-id="Exercises-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Exercises</a></div><div class="lev4 toc-item"><a href="#Hello-World---variable" data-toc-modified-id="Hello-World---variable-2.4.0.1"><span class="toc-item-num">2.4.0.1&nbsp;&nbsp;</span>Hello World - variable</a></div><div class="lev4 toc-item"><a href="#One-Variable,-Two-Messages:" data-toc-modified-id="One-Variable,-Two-Messages:-2.4.0.2"><span class="toc-item-num">2.4.0.2&nbsp;&nbsp;</span>One Variable, Two Messages:</a></div><div class="lev1 toc-item"><a href="#Overview-of-the-most-basic-data-types" data-toc-modified-id="Overview-of-the-most-basic-data-types-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Overview of the most basic data types</a></div><div class="lev1 toc-item"><a href="#Strings" data-toc-modified-id="Strings-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Strings</a></div><div class="lev2 toc-item"><a href="#Single-and-double-quotes" data-toc-modified-id="Single-and-double-quotes-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Single and double quotes</a></div><div class="lev2 toc-item"><a href="#Changing-case" data-toc-modified-id="Changing-case-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Changing case</a></div><div class="lev2 toc-item"><a href="#Combining-strings-(concatenation)" data-toc-modified-id="Combining-strings-(concatenation)-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Combining strings (concatenation)</a></div><div class="lev2 toc-item"><a href="#Whitespace" data-toc-modified-id="Whitespace-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Whitespace</a></div><div class="lev3 toc-item"><a href="#Stripping-whitespace" data-toc-modified-id="Stripping-whitespace-4.4.1"><span class="toc-item-num">4.4.1&nbsp;&nbsp;</span>Stripping whitespace</a></div><div class="lev2 toc-item"><a href="#More-string-functions" data-toc-modified-id="More-string-functions-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>More string functions</a></div><div class="lev2 toc-item"><a href="#Exercises" data-toc-modified-id="Exercises-4.6"><span class="toc-item-num">4.6&nbsp;&nbsp;</span>Exercises</a></div><div class="lev4 toc-item"><a href="#Someone-Said" data-toc-modified-id="Someone-Said-4.6.0.1"><span class="toc-item-num">4.6.0.1&nbsp;&nbsp;</span>Someone Said</a></div><div class="lev4 toc-item"><a href="#First-Name-Cases" data-toc-modified-id="First-Name-Cases-4.6.0.2"><span class="toc-item-num">4.6.0.2&nbsp;&nbsp;</span>First Name Cases</a></div><div class="lev4 toc-item"><a href="#Full-Name" data-toc-modified-id="Full-Name-4.6.0.3"><span class="toc-item-num">4.6.0.3&nbsp;&nbsp;</span>Full Name</a></div><div class="lev4 toc-item"><a href="#About-This-Person" data-toc-modified-id="About-This-Person-4.6.0.4"><span class="toc-item-num">4.6.0.4&nbsp;&nbsp;</span>About This Person</a></div><div class="lev4 toc-item"><a href="#Name-Strip" data-toc-modified-id="Name-Strip-4.6.0.5"><span class="toc-item-num">4.6.0.5&nbsp;&nbsp;</span>Name Strip</a></div><div class="lev1 toc-item"><a href="#Numbers" data-toc-modified-id="Numbers-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Numbers</a></div><div class="lev2 toc-item"><a href="#Integers" data-toc-modified-id="Integers-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Integers</a></div><div class="lev2 toc-item"><a href="#Floating-Point-numbers" data-toc-modified-id="Floating-Point-numbers-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Floating-Point numbers</a></div><div class="lev2 toc-item"><a href="#Exercises" data-toc-modified-id="Exercises-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Exercises</a></div><div class="lev4 toc-item"><a href="#Arithmetic" data-toc-modified-id="Arithmetic-5.3.0.1"><span class="toc-item-num">5.3.0.1&nbsp;&nbsp;</span>Arithmetic</a></div><div class="lev4 toc-item"><a href="#Order-of-Operations" data-toc-modified-id="Order-of-Operations-5.3.0.2"><span class="toc-item-num">5.3.0.2&nbsp;&nbsp;</span>Order of Operations</a></div><div class="lev4 toc-item"><a href="#Long-Decimals" data-toc-modified-id="Long-Decimals-5.3.0.3"><span class="toc-item-num">5.3.0.3&nbsp;&nbsp;</span>Long Decimals</a></div><div class="lev1 toc-item"><a href="#Jupyter's-'print',-operators-and-simple-function-calls" data-toc-modified-id="Jupyter's-'print',-operators-and-simple-function-calls-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Jupyter's 'print', operators and simple function calls</a></div><div class="lev1 toc-item"><a href="#Conversion-between-types" data-toc-modified-id="Conversion-between-types-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Conversion between types</a></div><div class="lev1 toc-item"><a href="#Comments" data-toc-modified-id="Comments-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Comments</a></div><div class="lev2 toc-item"><a href="#What-makes-a-good-comment?" data-toc-modified-id="What-makes-a-good-comment?-8.1"><span class="toc-item-num">8.1&nbsp;&nbsp;</span>What makes a good comment?</a></div><div class="lev2 toc-item"><a href="#When-should-you-write-comments?" data-toc-modified-id="When-should-you-write-comments?-8.2"><span class="toc-item-num">8.2&nbsp;&nbsp;</span>When should you write comments?</a></div><div class="lev2 toc-item"><a href="#Exercises" data-toc-modified-id="Exercises-8.3"><span class="toc-item-num">8.3&nbsp;&nbsp;</span>Exercises</a></div><div class="lev4 toc-item"><a href="#First-Comments" data-toc-modified-id="First-Comments-8.3.0.1"><span class="toc-item-num">8.3.0.1&nbsp;&nbsp;</span>First Comments</a></div><div class="lev1 toc-item"><a href="#OPTIONAL:-Exploring-the-Python-Community" data-toc-modified-id="OPTIONAL:-Exploring-the-Python-Community-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>OPTIONAL: Exploring the Python Community</a></div><div class="lev1 toc-item"><a href="#License" data-toc-modified-id="License-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>License</a></div>

Variables
===
A variable holds a value.

Example
---

In [1]:
message = "Hello Python world!"
print(message)

Hello Python world!


A variable holds a value. You can change the value of a variable at any point ('reassign' the variable).

In [2]:
message = "Hello Python world!"
print(message)

message = "Python is my favorite language!"
print(message)

Hello Python world!
Python is my favorite language!


Naming rules
---
- Variables can only contain letters, numbers, and underscores. Variable names can start with a letter or an underscore, but can not start with a number.
- Spaces are not allowed in variable names, so we use underscores instead of spaces. For example, use student_name instead of "student name".
- You cannot use [Python keywords](http://docs.python.org/3/reference/lexical_analysis.html#keywords) as variable names.
- Variable names should be descriptive, without being too long. For example mc_wheels is better than just "wheels", and number_of_wheels_on_a_motorycle.
- Be careful about using the lowercase letter l and the uppercase letter O in places where they could be confused with the numbers 1 and 0.

NameError
---
There is one common error when using variables, that you will almost certainly encounter at some point. Take a look at this code, and see if you can figure out why it causes an error.

In [6]:
message = "Thank you for sharing Python with the world, Guido!"
print(mesage)

NameError: name 'mesage' is not defined

Let's look through this error message. First, we see it is a NameError. Then we see the file that caused the error, and a green arrow shows us what line in that file caused the error. Then we get some more specific feedback, that "name 'mesage' is not defined".

You may have already spotted the source of the error. We spelled message two different ways. Python does not care whether we use the variable name "message" or "mesage". Python only cares that the spellings of our variable names match every time we use them.

This is pretty important, because it allows us to have a variable "name" with a single name in it, and then another variable "names" with a bunch of names in it.

We can fix NameErrors by making sure all of our variable names are spelled consistently.

In [7]:
message = "Thank you for sharing Python with the world, Guido!"
print(message)

Thank you for sharing Python with the world, Guido!


In case you didn't know [Guido](http://en.wikipedia.org/wiki/Guido_van_Rossum) [van Rossum](http://www.python.org/~guido/) created the Python language over 20 years ago, and he is considered Python's [Benevolent Dictator for Life](http://en.wikipedia.org/wiki/Benevolent_Dictator_for_Life). Guido still signs off on all major changes to the core Python language.

<a id="Exercises-variables"></a>
Exercises
---
#### Hello World - variable

- Store your own version of the message "Hello World" in a variable, and print it.

#### One Variable, Two Messages:
- Store a message in a variable, and then print that message.
- Store a new message in the same variable, and then print that new message.

[top](#)

# Overview of the most basic data types

| type | description |
|----|---|
| int | integer number |
| float  | floating point number |
| bool | True or False |
| string | text |

We can check the type of a variable with the "type" function.

In [9]:
print(type(5))
print(type(5.0))
print(type("Hello World"))
print(type('Hello World'))  # " " and ' ' can be used equivalently

<class 'int'>
<class 'float'>
<class 'str'>
<class 'str'>


Strings
===
Strings are sets of characters. Strings are easier to understand by looking at some examples.

Single and double quotes
---
Strings are contained by either single or double quotes.

In [10]:
my_string = "This is a double-quoted string."
my_string = 'This is a single-quoted string.'

This lets us make strings that contain quotations.

In [11]:
quote = "Linus Torvalds once said, 'Any program is only as good as it is useful.'"

Changing case
---
You can easily change the case of a string, to present it the way you want it to look.

In [12]:
first_name = 'eric'

print(first_name)
print(first_name.title())

eric
Eric


It is often good to store data in lower case, and then change the case as you want to for presentation. This catches some TYpos. It also makes sure that 'eric', 'Eric', and 'ERIC' are not considered three different people.

Some of the most common cases are lower, title, and upper.

In [13]:
first_name = 'eric'

print(first_name)
print(first_name.title())
print(first_name.upper())

first_name = 'Eric'
print(first_name.lower())

eric
Eric
ERIC
eric


You will see this syntax quite often, where a variable name is followed by a dot and then the name of an action, followed by a set of parentheses. The parentheses may be empty, or they may contain some values.

variable_name.action()

In this example, the word "action" is the name of a method. A method is something that can be done to a variable. The methods 'lower', 'title', and 'upper' are all functions that have been written into the Python language, which do something to strings. Later on, you will learn to write your own methods.

Combining strings (concatenation)
---
It is often very useful to be able to combine strings into a message or page element that we want to display. Again, this is easier to understand through an example.

In [14]:
first_name = 'ada'
last_name = 'lovelace'

full_name = first_name + ' ' + last_name

print(full_name.title())

Ada Lovelace


The plus sign combines two strings into one, which is called "concatenation". You can use as many plus signs as you want in composing messages. In fact, many web pages are written as giant strings which are put together through a long series of string concatenations.

In [15]:

first_name = 'ada'
last_name = 'lovelace'
full_name = first_name + ' ' + last_name

message = full_name.title() + ' ' + "was considered the world's first computer programmer."

print(message)

Ada Lovelace was considered the world's first computer programmer.


If you don't know who Ada Lovelace is, you might want to go read what [Wikipedia](http://en.wikipedia.org/wiki/Ada_Lovelace) or the [Computer History Museum](http://www.computerhistory.org/babbage/adalovelace/) have to say about her. Her life and her work are also the inspiration for the [Ada Initiative](http://adainitiative.org/faq/about-ada-lovelace/), which supports women who are involved in technical fields.

Whitespace
---
The term "whitespace" refers to characters that the computer is aware of, but are invisible to readers. The most common whitespace characters are spaces, tabs, and newlines.

Spaces are easy to create, because you have been using them as long as you have been using computers. Tabs and newlines are represented by special character combinations.

The two-character combination "\t" makes a tab appear in a string. Tabs can be used anywhere you like in a string.

In [16]:
print("Hello everyone!")

Hello everyone!


In [17]:
print("\tHello everyone!")

	Hello everyone!


In [18]:
print("Hello \teveryone!")

Hello 	everyone!


The combination "\n" makes a newline appear in a string. You can use newlines anywhere you like in a string.

In [19]:
print("Hello everyone!")

Hello everyone!


In [20]:
print("\nHello everyone!")


Hello everyone!


In [21]:
print("Hello \neveryone!")

Hello 
everyone!


In [22]:
print("\n\n\nHello everyone!")




Hello everyone!


### Stripping whitespace

Many times you will allow users to enter text into a box, and then you will read that text and use it. It is really easy for people to include extra whitespace at the beginning or end of their text. Whitespace includes spaces, tabs, and newlines.

It is often a good idea to strip this whitespace from strings before you start working with them. For example, you might want to let people log in, and you probably want to treat 'eric ' as 'eric' when you are trying to see if I exist on your system.

You can strip whitespace from the left side, the right side, or both sides of a string.

In [23]:
name = ' eric '

print(name.lstrip())
print(name.rstrip())
print(name.strip())

eric 
 eric
eric


It's hard to see exactly what is happening, so maybe the following will make it a little more clear:

In [24]:
name = ' eric '

print('-' + name.lstrip() + '-')
print('-' + name.rstrip() + '-')
print('-' + name.strip() + '-')

-eric -
- eric-
-eric-


## More string functions

In [25]:
s = "Python is my new favorite language"
len(s) # 34 characters long

34

In [26]:
s.startswith('P')

True

In [27]:
s.startswith('p')  # Be careful, its case sensitive

False

In [28]:
s.replace("my", "your")

'Python is your new favorite language'

In [29]:
s.replace("a", "4")  # replaces all occurences

'Python is my new f4vorite l4ngu4ge'

In [30]:
print(s) # but be careful: replace() (like the other functions) does not change the variable holding the string until you reassign 's', it just prints the result of the operation!

Python is my new favorite language


In [31]:
s = s.replace("a", "4")

In [32]:
print(s)

Python is my new f4vorite l4ngu4ge


In [33]:
s.replace("4", "a")
s.replace("o", "0") # really try to understand what is happening here!  1. replace gets called on 's', and replaces all 4s with a's, but does not reassign 's'. 
#2. Now 's' is called again, but the last replace is forgotten, so the end-result is a replace of o's with 0's, and the print-out of that. 's' is still the same. 

'Pyth0n is my new f4v0rite l4ngu4ge'

In [34]:
s.replace("a", "4").replace("o", "0")

'Pyth0n is my new f4v0rite l4ngu4ge'

In [35]:
s.split(" ")

['Python', 'is', 'my', 'new', 'f4vorite', 'l4ngu4ge']

We can split strings at defined characters

In [36]:
message = "I like cats and dogs, but I'd much rather own a dog."
words = message.split(' ') #split at whitespaces
print(words)

['I', 'like', 'cats', 'and', 'dogs,', 'but', "I'd", 'much', 'rather', 'own', 'a', 'dog.']


Notice that the punctuation is left in the substrings.

It is more common to split strings that are really lists, separated by something like a comma. The *split()* function gives you an easy way to turn comma-separated strings, which you can't do much with in Python, into lists. Once you have your data in a list, you can work with it in much more powerful ways.

In [37]:
animals = "dog, cat, tiger, mouse, liger, bear"

# Rewrite the string as a list, and store it in the same variable
animals = animals.split(',')
print(animals)

['dog', ' cat', ' tiger', ' mouse', ' liger', ' bear']


Notice that in this case, the spaces are also ignored. It is a good idea to test the output of the split() function and make sure it is doing what you want with the data you are interested in.
One use of this is to work with spreadsheet data in your Python programs. Most spreadsheet applications allow you to dump your data into a comma-separated text file. You can read this file into your Python program, or even copy and paste from the text file into your program file, and then turn the data into a list. You can then process your spreadsheet data using a for loop.

<a id="Exercises-strings"></a>
Exercises
---
#### Someone Said
- Find a quote that you like. Store the quote in a variable, with an appropriate introduction such as "Ken Thompson once said, 'One of my most productive days was throwing away 1000 lines of code'". Print the quote.

#### First Name Cases
- Store your first name, in lowercase, in a variable.
- Using that one variable, print your name in lowercase, Titlecase, and UPPERCASE.

#### Full Name
- Store your first name and last name in separate variables, and then combine them to print out your full name.

#### About This Person
- Choose a person you look up to. Store their first and last names in separate variables.
- Use concatenation to make a sentence about this person, and store that sentence in a variable.-
- Print the sentence.

#### Name Strip
- Store your first name in a variable, but include at least two kinds of whitespace on each side of your name.
- Print your name as it is stored.
- Print your name with whitespace stripped from the left side, then from the right side, then from both sides.

[top](#)

Numbers
===
Dealing with simple numerical data is fairly straightforward in Python, but there are a few things you should know about.

Some Arithmetic Operators:

| Symbol | Task Performed |
|----|---|
| +  | Addition |
| -  | Subtraction |
| /  | division |
| //  | integer division |
| *  | multiplication |
| **  | to the power of |



Integers
---
You can do all of the basic operations with integers, and everything should behave as you expect. Addition and subtraction use the standard plus and minus symbols. Multiplication uses the asterisk, and division uses a forward slash. Exponents use two asterisks.

In [38]:
print(3+2)

5


In [39]:
print(3-2)

1


In [40]:
print(3*2)

6


In [41]:
print(3/2)

1.5


In [42]:
print(3**2)

9


----------------------

You can use parentheses to modify the standard order of operations.

In [43]:
standard_order = 2+3*4
print(standard_order)

14


In [44]:
my_order = (2+3)*4
print(my_order)

20


In [45]:
print(my_order + standard_order)

34


In [46]:
print(my_order / my_order)

1.0


Floating-Point numbers
---
Floating-point numbers refer to any number with a decimal point. Most of the time, you can think of floating point numbers as decimals, and they will behave as you expect them to.

In [47]:
print(0.1+0.1)

0.2


However, sometimes you will get an answer with an unexpectly long decimal part:

In [61]:
print(0.1+0.2)

0.30000000000000004


This happens because of the way computers represent numbers internally; this has nothing to do with Python itself. Basically, we are used to working in powers of ten, where one tenth plus two tenths is just three tenths. But computers work in powers of two. So your computer has to represent 0.1 in a power of two, and then 0.2 as a power of two, and express their sum as a power of two. There is no exact representation for 0.3 in powers of two, and we see that in the answer to 0.1+0.2.

Python tries to hide this kind of stuff when possible. Don't worry about it much for now; just don't be surprised by it, and know that we will learn to clean up our results a little later on.

You can also get the same kind of result with other operations.

In [62]:
print(3*0.1)

0.30000000000000004


----------------------

Be sure you know what number type you get out of different operations.

In [63]:
print(3/2)

1.5


In [64]:
print(3//2)

1


If you are getting numerical results that you don't expect for division, or that don't make sense, check if the version of Python you are using is treating integers differently than you expect.
**Our class will only use Python 3.0 and above!**

<a id="Exercises-numbers"></a>
Exercises
---
#### Arithmetic
- Write a program that prints out the results of at least one calculation for each of the basic operations: addition, subtraction, multiplication, division, and exponents.

#### Order of Operations
- Find a calculation whose result depends on the order of operations.
- Print the result of this calculation using the standard order of operations.
- Use parentheses to force a nonstandard order of operations. Print the result of this calculation.

#### Long Decimals
- On paper, 0.1+0.2=0.3. But you have seen that in Python, 0.1+0.2=0.30000000000000004.
- Find at least one other calculation that results in a long decimal like this.



# Jupyter's 'print', operators and simple function calls



Conveniently, in Jupyter the last statement of a cell gets printed automatically. Typically, in other Python interpreters/IDEs you have to use the print function instead, like we did above and will do in most of the remainder of this tutorial. 

In [65]:
2 + 2

4

In [66]:
2 + 2
4 + 5
'random string I just thought of'

'random string I just thought of'

Using 'print' explictly allows you multiple print-outs per cell.

In [67]:
print(2 + 2)
print(4 + 5)
print('random string I just thought of')

4
9
random string I just thought of


----------------------

Python evaluates if a given statement is either True or False and return the corresponding boolean value. 

Some relational operators:

| Symbol | Task Performed |
|----|---|
| == | True, if both sides are equal |
| !=  | True, if both sides are not equal |
| < |True, if the left side is smaller |
| > | True, if the left side is greater |
| <=  | True, if the left side is smaller or equal to the right side |
| >=  | True, if the left side is greater or equal to the right side |

In [68]:
5 == 3  #The difference to a single '=' is crucial!

False

In [69]:
3 > 3

False

In [70]:
3 >= 3

True

In [71]:
standard_order > 5 #we defined 'standard_order above! Remember that all variables get stored once assigned until you shutdown your notebook or overwrite them. 

True

In [72]:
standard_order != 66

True

In [73]:
6 == 'a random string'

False

In [48]:
6 == a_variable_we_did_not_define_yet

NameError: name 'a_variable_we_did_not_define_yet' is not defined

----------------------

There are several built-in function in Python that you can always call, see https://docs.python.org/3/library/functions.html for a full list

In [49]:
abs(-5)

5

In [50]:
max(3, 5)

5

In [51]:
max(x, 5)

NameError: name 'x' is not defined

----------------------

We can use other types in a similar way as numbers, quite intuitively:

In [78]:
'A' < 'B'

True

In [79]:
max("abc", "zyx")

'zyx'

In [80]:
max("abc", "8") 

'abc'

In [81]:
max("abc", 8) # mixing types doesn't work

TypeError: '>' not supported between instances of 'int' and 'str'

In [82]:
max("abc", str(8)) # so we have to transform (see below)

'abc'

----------------------

And of course, different data types allow for different functions:

In [83]:
len("How long is this string?")

24

In [84]:
x = 10
len(x)

TypeError: object of type 'int' has no len()

# Conversion between types
You can convert between these different by casting:

In [85]:
x = 10

print(type(x))
print(x)
x_as_string = str(x)

print(type(x_as_string))
print(x_as_string)
print(len(x_as_string))

<class 'int'>
10
<class 'str'>
10
2


----------------------

A typical conversion for prints:

In [86]:
print("My variable x is " + str(x))

My variable x is 10


[top](#)

Comments
===
As you begin to write more complicated code, you will have to spend more time thinking about how to code solutions to the problems you want to solve. Once you come up with an idea, you will spend a fair amount of time troubleshooting your code, and revising your overall approach.

Comments allow you to write in English, within your program. In Python, any line that starts with a pound (#) symbol is ignored by the Python interpreter.

In [87]:
# This line is a comment.
print("This line is not a comment, it is code.")

This line is not a comment, it is code.


What makes a good comment?
---
- It is short and to the point, but a complete thought. Most comments should be written in complete sentences.
- It explains your thinking, so that when you return to the code later you will understand how you were approaching the problem.
- It explains your thinking, so that others who work with your code will understand your overall approach to a problem.
- It explains particularly difficult sections of code in detail.

When should you write comments?
---
- When you have to think about code before writing it.
- When you are likely to forget later exactly how you were approaching a problem.
- When there is more than one way to solve a problem.
- When others are unlikely to anticipate your way of thinking about a problem.

Writing good comments is one of the clear signs of a good programmer. If you have any real interest in taking programming seriously, start using comments now. You will see them throughout the examples in these notebooks.

<a id="Exercises-comments"></a>
Exercises
---
#### First Comments
- Try commenting in different places in the code that we have written so far and explore what happens when you execute


[top](#)

    
# OPTIONAL: Exploring the Python Community

The Python community is incredibly rich and diverse. Here are a couple resources to look at, if you want to do some exploring.

- [The Python website](http://python.org/)

    The main Python website is probably not of too much interest to you at this point, but it is a great resource to know about as you start to learn more.
    
<!-- -->

- [PyCon](https://us.pycon.org/)

    The Python Conference (PyCon) is an incredible event, and the community is entirely welcoming to new programmers. They happen all over the world, throughout the year. If you can make your way to one of these conferences, you will learn a great deal and meet some really interesting people.

<!-- -->
    
- [PyLadies](http://www.pyladies.com/)

    Women and minorities are still under-represented in most technology fields, and the programming world is no different in this regard. That said, the Python community may well be the most welcoming and supportive programming community for women and minorities. There are a number of groups dedicated to bringing women and minorities together around programming in Python, and there are a number of explicit Codes of Conduct for Python-related events.

    PyLadies is one of the most visible of these organizations. They are a great resource, so go see what they do and what they have to offer.

<!-- -->
    
- [Python User Groups](https://wiki.python.org/moin/LocalUserGroups)

    Wherever there are a number of Python programmers, they will find a way to get together. Python user groups are regular meetings of Python users from a local area. Go take a look at the list of user groups, and see if there is one near you.

[top](#)

# License

The MIT License (MIT)

Original work Copyright (c) 2013 Eric Matthes  
Modified work Copyright 2017 Fabian Flöck, Florian Lemmerich  

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.